### PR TITLE
RFC 213: UI changes on sign-in/sign-up form

### PR DESCRIPTION
**Pull request by tjkandala**
### Sign up GIFs

Implementation: Add two public endpoints to check for username and email uniqueness. Implement `createValidationPipeline` to generalize async + sync validation logic for our `useEventObservable` hook.

<details>
<summary>Synchronous validation (no loading spinner when errors found client-side)
</summary>

</details>

<details>
<summary>Asynchronous validation</summary>

</details>

<details open>
<summary>Username validation (both asynchronous and synchronous)</summary>

</details>

<details>
<summary>Password validation (always synchronous)</summary>

</details>

<details>
<summary>Register</summary>

</details>

### Third party auth errors

Implementation: Instead of calling `http.Error`, redirect to `/sign-in` with the error message as a URL parameter. The client displays an error message + strips URL parameters.

Screenshot from 2020-09-21 19-50-25

TODO for RFC 213
- [x] - Remove sign out page (separate PR)
- [ ] - Use new redirect func for all errors (implemented on some errors for review)

### CODE REVIEW

**Comment by quinnkeast** [comment]
Looking great!

Here's some remarks as I go. (If something's outside of review scope for now, I'll just save it for later!)

### Draft

**Visual**
- [ ] Border colour on the container element on light mode appears to be `border 1` – ought be `border 2`, a bit lighter.
- [ ] Error message text should be left-aligned rather than centred.
- [ ] On the sign up view in dark mode, the `At least 12 characters` helper ought be `text 2` (a bit brighter than the labels).
- [ ] On the sign in to server view with only username/password, there's extra margin on the `form-group` containing the Sign in button, which adds extra space between the button and the bottom of the container (should be `1.5rem` total).

**Copy**
- [ ] On sign in view for server, `To create an account, contact the site admin.` should become `Need an account? Contact your site admin` (no period at end of line).
- [ ] On sign up view, `Already have an account? Sign in.` should become `Already have an account? Sign in` (no period at end of line).
- [ ] **Question** – where is the copy for the inline error messages kept? I'd like to make some tweaks to these, but thinking it might be easier if I did it directly rather than providing it here...

**Interaction**
- [ ] Synchronous validation should only trigger when the input is blurred. In other words, an invalid email address wouldn't show an error until the user leaves the field. Otherwise, users are told "they're doing it wrong" while they're in the process of filling out the field.

**Review by felixfbecker — COMMENTED** [review]
Really impressed by the Rx validation pipeline and that you figured out the backend, awesome work!

File: web/src/auth/SignUpForm.tsx [hunk 1]
```diff
@@ -28,66 +36,281 @@ interface SignUpFormProps {
...
+     * return with the reason the input is invalid.
+     */
+    synchronousValidators?: ((value: string) => string | void)[]
+
+    /**
+     * Optional array of asynchronous input validators. These must return
+     * observables created with `fromFetch` for easy cancellation in `switchMap`.
+     *
+     * If there's no problem with the input, void return. Else,
+     * return with the reason the input is invalid.
```
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignUpForm.tsx:56]: oops, these used to be promises. good catch!

File: web/src/auth/SignUpForm.tsx [hunk 2]
```diff
@@ -28,66 +36,281 @@ interface SignUpFormProps {
...
+    )
 
-    public render(): JSX.Element | null {
-        return (
+    const [emailState, setEmailState] = useState({ value: '', loading: false })
+    const [nextEmailFieldChange, emailValidationResult] = useEventObservable<
+        React.ChangeEvent<HTMLInputElement>,
+        ValidationResult
+    >(
+        useMemo(() => createValidationPipeline('email', setEmailState, signUpFieldValidators.email), [
```
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignUpForm.tsx:141]: @quinnkeast expressed that synchronous validation should only trigger when the input is blurred. In our existing forms, we only give immediate feedback after the first attempt. How should we coordinate that with asynchronous validation? 

A user could input an invalid email, confirm that it is unique, then find out that it's invalid after submitting. I propose that we keep the current behavior, but debounce synchronous validation as well. Alternatively, if we want to conform to our existing forms, asynchronous validation would provide no added value, so we'd rip that out.

File: web/src/auth/SignInSignUpPageCommon.scss [hunk 3]
```diff
@@ -9,6 +9,29 @@
...
+    border-color: var(--color-border-2);
+
+    &__input-container {
+        position: relative;
+    }
+
+    &__icon {
+        position: absolute;
+        right: 0.4375rem;
+        top: 0.4375rem;
```
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignInSignUpPageCommon.scss:23]: It is 0.5rem in the designs, but the input container height is 34px instead of 32px to account for the 1px border. To center the logo, I had to position it 7px from the top.

[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/auth/SignInSignUpPageCommon.scss:23]: Ah - maybe make it `calc(0.5rem - 1px)` for clarity?

File: web/src/auth/SignUpForm.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (quinnkeast) [reply web/src/auth/SignUpForm.tsx:141]: Apologies here: I don't think I explained my thoughts clearly enough.

- Synchronous validation should be done on blur for **formatting requirements**, e.g. valid email format or password length.
- Submit validation should be done for **server-side validation** like email uniqueness.
- The one exception to the above is the username in this signup flow, which we're validating synchronously for availability.

This pattern means:

- Users don't have to submit first, before recovering from simple formatting issues.
- Users don't get errors while they're in the process of typing the first time.
- Server-side validation is separate from formatting validation.

In this context, additional positive feedback when server-side validation fails isn't necessary.

File: web/src/auth/SignInPage.tsx [hunk 4]
```diff
@@ -18,13 +24,69 @@ interface SignInPageProps {
 export const SignInPage: React.FunctionComponent<SignInPageProps> = props => {
     useEffect(() => eventLogger.logViewEvent('SignIn', null, false))
 
+    const [authError, setAuthError] = useState<Error | null>(() => {
+        // Display 3rd party auth errors (redirect with param 'auth_error')
+        const authErrorMessage = new URLSearchParams(location.search).get('auth_error')
```
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignInPage.tsx:29]: I think it's for errors _dealing_ with 3rd parties, but I'll find a better name. Regarding the hardcoded errors, should I include the most common errors, then display "Ask your site admin to check logs" for unknown/low-priority codes?

File: web/src/global-styles/web-content.scss [hunk 5]
```diff
@@ -10,6 +10,7 @@
     }
     .theme-light & {
         --link-color: #1c7cd6;
+        --body-color: #{$gray-18};
```
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/global-styles/web-content.scss:13]: Screenshot from 2020-09-22 17-38-35
I saw that this color was specified as body-color light [here](https://www.figma.com/file/L5ZdNBGAnHQQUGGmp6tdQS/%2312695-%E2%80%93-Sign-in-%2F-Out-UX-%5BApproved%5D?node-id=17%3A2040) so I made a faulty assumption. I'll set color to gray-18 only where needed instead

File: web/src/auth/SignUpForm.tsx [hunk 6]
```diff
@@ -28,66 +36,281 @@ interface SignUpFormProps {
...
+                            })}
+                            onChange={nextEmailFieldChange}
+                            required={true}
+                            value={emailState.value}
+                            disabled={loading}
+                            autoFocus={true}
+                            placeholder=" "
+                        />
+                        {emailState.loading ? (
+                            <LoadingSpinner className="signin-signup-form__icon" />
```
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignUpForm.tsx:237]: Yes, and should I make a similar component for btn-block with a loader, or is that simple enough to not componentize?

File: web/src/auth/SignUpForm.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignUpForm.tsx:141]: Thanks @quinnkeast! I was under the impression that [emails and usernames should be handled the same way](https://www.figma.com/file/L5ZdNBGAnHQQUGGmp6tdQS/%2312695-%E2%80%93-Sign-in-%2F-Out-UX-%5BApproved%5D?node-id=17%3A1798).

@felixfbecker I think we can use the same pipeline with synchronous validation on blur, I'd just need to rework it to take two types of events and cancel network requests when an `onblur` observable emits formatting errors. My concern is checking for availability when the formatting is wrong. It might be misleading to suggest that an invalid username or email is available.

[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignUpForm.tsx:201]: Update: New behavior for form. It only checks for sync (formatting) validation errors on change after there has been a sync validation error on blur before. The implementation complexity has increased a bit, so I'll look into either marble or integration testing the validation pipeline.

**Comment by tjkandala** [comment]
> * [ ]  **Question** – where is the copy for the inline error messages kept? I'd like to make some tweaks to these, but thinking it might be easier if I did it directly rather than providing it here...

The copy comes from the strings returned from [these validation functions](https://github.com/sourcegraph/sourcegraph/pull/13939/files#diff-ee0c7e6ecae49f0c7ec09f25e7741514R438). The implementation may change over time, but it won't be hard to keep the same copy, so you can go ahead and tweak the messages

File: web/src/auth/SignUpForm.tsx [hunk 7]
```diff
@@ -28,66 +37,343 @@ interface SignUpFormProps {
...
+                    )
+                }
+
+                return zip(...asynchronousValidators.map(validator => validator(validationEvent.value)))
+                    .pipe(
+                        map(values => head(compact(values))),
+                        map(reason => (reason ? { kind: 'INVALID' as const, reason } : { kind: 'VALID' as const })),
+                        tap(() => onInputChange(previousInputState => ({ ...previousInputState, loading: false })))
+                    )
+                    .pipe(takeUntil(syncReasons))
```
Comment (tjkandala) [inline web/src/auth/SignUpForm.tsx:164]: TODO: This pipe now seems to be redundant

File: web/src/auth/SignUpForm.tsx [hunk 8]
```diff
@@ -28,66 +36,281 @@ interface SignUpFormProps {
...
+                            'text-danger font-weight-bold':
+                                emailValidationResult?.kind === 'INVALID' && !emailState.loading,
+                        })}
+                    >
+                        Email
+                    </label>
+                    <div className="signin-signup-form__input-container">
+                        <EmailInput
+                            className={classNames('signin-signup-form__input', {
+                                'border-danger': emailValidationResult?.kind === 'INVALID' && !emailState.loading,
```
[the remark this answers is not in the document]
Comment (quinnkeast) [reply web/src/auth/SignUpForm.tsx:287]: I feel fairly strongly about it for a few reasons:

- The native error messages are temporary. Once you start typing in a field to correct it, the message itself disappears.
- The native error messages only show one at a time. If there's multiple errors, it's a good pattern to reveal all of those errors in one go.
- The native error messages are sometimes unreliable. For example, I use 1Password; when I submit a signup form, 1Password triggers a system dialogue, which "cancels out" the native error messages, and I don't see any feedback. 
- Using a custom approach to how error messages are shown vs. native error messages opens up more opportunity to make the UI design our own, gives us greater control over accessibility, and so on.

>If yes, I think there is a way to prevent the native UI and replace it with custom rendering1, but we should probably do that for all forms then, else it would be inconsistent (and that should probably be a separate PR).

I'd be in favour of moving in that direction, but also agree that should be a separate PR, as I foresee ripple effects... I'd want to take a look throughout the platform for all cases where errors might appear and make sure nothing would break as a result.

File: web/src/auth/SignUpForm.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (quinnkeast) [reply web/src/auth/SignUpForm.tsx:201]: >Thanks @quinnkeast! I was under the impression that emails and usernames should be handled the same way.

This is great feedback for me to make sure I clarify in Figma in the future!

File: web/src/auth/SignInSignUpPageCommon.scss [hunk 9]
```diff
@@ -9,6 +9,29 @@
 }
 
 .signin-signup-form {
-    width: 22rem;
     max-width: calc(100vw - 2rem);
+    background-color: var(--body-bg-1);
+    border-color: var(--color-border-2);
```
[the remark this answers is not in the document]
Comment (quinnkeast) [reply web/src/auth/SignInSignUpPageCommon.scss:14]: Yes – it's subtle, but the default border colour for the `card` is too dark for this context. (Separate concern: I think the card itself might benefit from the lighter colour as well, but I don't want to make that sweeping change yet.)

File: web/src/auth/SignInPage.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/auth/SignInPage.tsx:29]: It doesn't seem feasible to hardcode/enumerate third party errors, sorry, I had only seen instances in the backend where `auth_error` came from our backend. How do we handle these situations today? Is there a secure, non-URL way we could transfer this message, maybe `window.context`? Maybe a cookie?

File: web/src/auth/SignUpForm.tsx (hunk 2 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/auth/SignUpForm.tsx:201]: We can use `reportValidity()` to trigger a synchronous validation on blur and kick off the async validation (which can then report validity with `setCustomValidity()`)

File: web/src/auth/SignUpForm.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/auth/SignUpForm.tsx:287]: Got it. In any case we need to change the code to use the native form validation API so that our CSS `:invalid` styles properly match.

Should we already do the work just for sign-in/sign-out to intercept the native UI and display in an ARIA Live region instead, or keep it with native UI for this PR and follow-up with changing it generally for all forms?

File: web/src/auth/SignUpForm.tsx (hunk 6 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/auth/SignUpForm.tsx:298]: Seems like a good idea if that is a general pattern we're gonna use to indicate loading after a CTA was clicked (we're currently inconsistent in that regard).
The `-block` part would just be a passed `className` though, so I'm thinking of just something like `LoaderButton` with the same props as `<button>`.

File: web/src/auth/SignUpForm.tsx (hunk 8 above)
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignUpForm.tsx:287]: I'd say option two based on our goal to merge by EOW, but I'm curious as to how that would work with the username uniqueness validation.

File: web/src/auth/SignInPage.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignInPage.tsx:29]: They're currently sent as plain text through [`http.Error`](https://golang.org/src/net/http/server.go?s=63178:63230#L2041).

File: cmd/frontend/auth/redirect.go [hunk 10]
```diff
@@ -23,3 +24,8 @@ func SafeRedirectURL(urlStr string) string {
 	u = &url.URL{Path: u.Path, RawQuery: u.RawQuery}
 	return u.String()
 }
+
+// Redirects to sign in page to display error messages after third-party auth errors
```
Comment (unknwon) [inline cmd/frontend/auth/redirect.go:28]:
```suggestion
// Redirects to sign in page to display error messages after third-party auth errors.
//
// 🚨 SECURITY: The `message` must not contain any confidential information.
```

File: enterprise/cmd/frontend/auth/openidconnect/config.go [hunk 11]
```diff
@@ -25,23 +25,23 @@ func getProvider(id string) *provider {
...
 
-func handleGetProvider(ctx context.Context, w http.ResponseWriter, id string) (p *provider, handled bool) {
+func handleGetProvider(ctx context.Context, w http.ResponseWriter, r *http.Request, id string) (p *provider, handled bool) {
 	handled = true // safer default
 
 	p = getProvider(id)
 	if p == nil {
 		log15.Error("No OpenID Connect auth provider found with ID.", "id", id)
-		http.Error(w, "Misconfigured OpenID Connect auth provider.", http.StatusInternalServerError)
+		http.Redirect(w, r, "/sign-in?auth_error=Misconfigured OpenID Connect auth provider.", http.StatusFound)
```
Comment (unknwon) [inline enterprise/cmd/frontend/auth/openidconnect/config.go:34]: Any reason we don't use `ProviderErrorRedirect` here and other places in this file?

File: web/src/auth/SignInPage.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/auth/SignInPage.tsx:29]: If we serve the response, can we just serve the sign-in page as a response, and put the error in `JSContext` for the client to read?

File: enterprise/cmd/frontend/auth/openidconnect/config.go (hunk 11 above)
Comment (tjkandala) [reply enterprise/cmd/frontend/auth/openidconnect/config.go:34]: Just wanted to make sure there weren't any glaring issues before replacing all of the provider errors. Thanks!

File: cmd/frontend/auth/redirect.go [hunk 12]
```diff
@@ -23,3 +24,10 @@ func SafeRedirectURL(urlStr string) string {
 	u = &url.URL{Path: u.Path, RawQuery: u.RawQuery}
 	return u.String()
 }
+
+// Redirects to sign in page to display error messages after third-party auth errors.
+//
+// 🚨 SECURITY: The `message` must not contain any confidential information.
+func ProviderErrorRedirect(w http.ResponseWriter, r *http.Request, message string) {
+	http.Redirect(w, r, "/sign-in?auth_error="+message, http.StatusFound)
```
Comment (unknwon) [inline cmd/frontend/auth/redirect.go:32]: Just noticed the `message` might contain spaces, have you had chance to verify it works in major browsers? Otherwise, we might need to escape spaces ourselves here:

```suggestion
	http.Redirect(w, r, "/sign-in?auth_error="+url.QueryEscape(message), http.StatusFound)
```

**Review by unknwon — APPROVED** [review]
Other than [the security concern raised by @felixfbecker](https://github.com/sourcegraph/sourcegraph/pull/13939/files#r492708280) and the [re-use of `ProviderErrorRedirect`](https://github.com/sourcegraph/sourcegraph/pull/13939/files#r494061088), LGTM. Approve to unblock.

File: web/src/auth/ResetPasswordPage.tsx [hunk 13]
```diff
@@ -39,42 +39,60 @@ class ResetPasswordInitForm extends React.PureComponent<ResetPasswordInitFormPro
...
-                    <Link to="/sign-in">Return to sign in</Link>
-                </small>
-                {this.state.submitOrError === 'loading' && <LoadingSpinner className="icon-inline mt-2" />}
+            <>
                 {isErrorLike(this.state.submitOrError) && (
                     <ErrorAlert className="mt-2" error={this.state.submitOrError} history={this.props.history} />
                 )}
-            </Form>
+                <Form
+                    className="reset-password-page__form signin-signup-form border rounded p-4 mb-3"
```
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/ResetPasswordPage.tsx:60]: Resolved:

> Yes – it's subtle, but the default border colour for the card is too dark for this context. (Separate concern: I think the card itself might benefit from the lighter colour as well, but I don't want to make that sweeping change yet.)

File: web/src/auth/SignInPage.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (tjkandala) [reply web/src/auth/SignInPage.tsx:29]: `serveSignIn` isn't accessible from the provider packages since it's in an internal package. Given the concern of phishing, we should revisit this later. Could we disable markdown rendering for some error messages?

**Comment by tjkandala** [comment]
Update: 
- Decided in pairing session with @felixfbecker to validate synchronously on input as well to avoid displaying a valid state ('Username is unique') when formatting errors are present. (input -> 300ms debounce -> synchronous checks -> asynchronous checks)
- Removed `ProviderErrorRedirect` for now as it enables phishing
- Ready to merge upon approval by @sourcegraph/security (concerns over [username enumeration](https://github.com/sourcegraph/sourcegraph/pull/13939#discussion_r492708280))

File: web/src/auth/SignInPage.tsx (hunk 4 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/auth/SignInPage.tsx]: Disabling markdown wouldn't help much, I could still make the message "Per new company policy you need to reset your password every 64 days. Please reset your password at https://sourcgraph.com/reset-password before you can sign in again." and users would likely not notice the difference.

[the remark this answers is not in the document]
Comment (felixfbecker) [reply web/src/auth/SignInPage.tsx]: I agree we should revisit this though. I don't know how the code would need to be restructured to make this work, maybe @sourcegraph/cloud would be better to own this issue?

File: cmd/frontend/internal/auth/userpasswd/handlers.go [hunk 14]
```diff
@@ -238,6 +240,39 @@ func HandleSignIn(w http.ResponseWriter, r *http.Request) {
 	}
 }
 
+// Check availability of username for signup form
+func HandleCheckUsernameTaken(w http.ResponseWriter, r *http.Request) {
```
Comment (eseliger) [inline cmd/frontend/internal/auth/userpasswd/handlers.go:244]: I think a username is still taken, when an org with that name exists, so this feels unsafe. Sqs's recently added API to lookup a namespace by name would do what you want I think: https://sourcegraph.com/github.com/sourcegraph/sourcegraph@main/-/blob/cmd/frontend/graphqlbackend/schema.graphql#L2280-2288

Comment (eseliger) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: cc @unknwon to confirm :)

Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: @eseliger is that API available to unauthenticated users?

Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: Just checked, appears to be: [API Console](https://sourcegraph.com/api/console#%7B%22query%22%3A%22query%20%7B%5Cn%20%20namespaceByName(name%3A%20%5C%22test%5C%22)%20%7B%5Cn%20%20%20%20__typename%5Cn%20%20%20%20...%20on%20Org%20%7B%5Cn%20%20%20%20%20%20name%5Cn%20%20%20%20%7D%5Cn%20%20%20%20...%20on%20User%20%7B%5Cn%20%20%20%20%20%20username%5Cn%20%20%20%20%7D%5Cn%20%20%7D%5Cn%7D%22%2C%22variables%22%3A%22%22%7D)

In that case there is also no security concern because we're already exposing this anyway. We should use that API, thanks @eseliger!

Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Ah nevermind, it's only accessible publicly for sourcegraph.com :|

Comment (eseliger) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: I see, might need to be exposed, if that's a feature we want to offer? At least, the logic should be taken from there since org names and user names live in the same "namespace"

Comment (unknwon) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: >I think a username is still taken, when an org with that name exists, so this feels unsafe.

Hmm, seems we're having separate tables for users and orgs, so their names shouldn't conflict to each other.

Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: I adapted the logic in https://github.com/sourcegraph/sourcegraph/pull/13939/commits/a12cf1acc4e754e894f940f3569db4bc909cf9c2

Comment (eseliger) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: That's what the `names` table is used for, no?

Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: @unknwon we enforce uniqueness regardless - I just did the reverse check, creating an org for an existing user:

image

Comment (unknwon) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: Gotcha. Thanks for the find!

File: cmd/frontend/internal/auth/userpasswd/handlers.go [hunk 15]
```diff
@@ -238,6 +240,39 @@ func HandleSignIn(w http.ResponseWriter, r *http.Request) {
...
+	if err != nil {
+		w.WriteHeader(http.StatusNotFound)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
+}
+
+// Check availability of email for signup form
+func HandleCheckEmailTaken(w http.ResponseWriter, r *http.Request) {
```
[the remark this answers is not in the document]
Comment (ElizabethStirling) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Yeah, that would be a security risk, but I wouldn't see it as being terribly high priority. Why do we need to expose this endpoint?

File: cmd/frontend/internal/auth/userpasswd/handlers.go (hunk 14 above)
Comment (ElizabethStirling) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: Yeah, definitely seems like something we should fix, since it would let anyone block certain customers from Sourcegraph cloud, since they could create an account with the same name as the organization they don't want us to be able to create.

Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go:244]: (note this was already fixed, resolving thread)

File: cmd/frontend/internal/auth/userpasswd/handlers.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Part of the RFC for sign-up UX improvements is showing immediate (on-type) validation of whether the username is available or already taken, vs showing an error alert after hitting enter (submitting the form).

[the remark this answers is not in the document]
Comment (ElizabethStirling) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Ahh, I see why rate limiting might be difficult then, if we're checking every time the user enters a character. If we could update it to only check after the user moves to a different field, that might make rate limiting feasible. Even limiting it to 1 check per 1-5 seconds, and maxing out at 200/ip/day could make this enumeration attack mostly infeasible.

[the remark this answers is not in the document]
Comment (ElizabethStirling) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Though, this would be super difficult if we consider corporate networks where many people might be creating accounts at the same time. @chayim do you have any thoughts on this? Not certain if this is a super significant issue, or something we'd need to be super wary of.

[the remark this answers is not in the document]
Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: We're already debouncing it so it won't check it until the user stops typing. General case we would just do one request if it comes back green, or as many needed until it's green, so just as many checks as before with validation on submit.

[the remark this answers is not in the document]
Comment (ElizabethStirling) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: That makes sense - rate limiting on successful requests seems like a potentially reasonable solution here.

[the remark this answers is not in the document]
Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Okay, so just confirming you'd want that to be implemented before merging?

[the remark this answers is not in the document]
Comment (ElizabethStirling) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: If it's a simple fix, then yes. I wouldn't lose sleep if we end up needing to push it off, though.

[the remark this answers is not in the document]
Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: It's not a simple fix for _us_ since we're all frontend engineers on the web team 😬 
Could we get some help on this?

[the remark this answers is not in the document]
Comment (ElizabethStirling) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Ahh, gotcha gotcha. I'll try to take a quick look before lunch, but I'm not immediately certain how I'd implement that rate limiting without causing issues for enterprise networks. Might see who implemented rate limiting somewhere else in the product for feedback in this thread, since they've probably thought more about the implementation than I have.

Tl;dr, I'll take a quick look, and get back to ya

[the remark this answers is not in the document]
Comment (ElizabethStirling) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Looks like this might be tricky. One solution would be to make our own rate limiter by storing hits in redis by the IP of the requester, but this would be a lot of work to implement and test. A quick hacky solution would be to just slow down each request, but that seems bad. I don't think this is high priority, and in order to not block this, I'm fine with merging this and just logging the issue in security-issues. Lmk if that works for you, and I'll create the issue.

[the remark this answers is not in the document]
Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Sounds good!

**Review by felixfbecker — APPROVED** [review]

**Review by quinnkeast — APPROVED** [review]
LGTM. Will follow up on copy edits as a separate issue!

File: cmd/frontend/internal/auth/userpasswd/handlers.go (hunk 15 above)
[the remark this answers is not in the document]
Comment (chayim) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: If this is for the UI only, is there any reason we can't implement something like the Reddit username creation flow. Yes, they add a captcha, yes captchas are fairly awful, but society has accepted them.

On the rate-limiting side I was experimenting with the following library in something I'd been looking at. There's always the standard caveat - people behind the same network IP (i.e an office address) have a real issue to contend with, which is why even Google ended up going all in on the captcha side.

https://github.com/didip/tollbooth

[the remark this answers is not in the document]
Comment (felixfbecker) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: Sorry, I don't understand what the Reddit flow would solve here, mind elaborating? It seems like Reddit also allows checking uniqueness of the username without filling out the chapta:

image

The chapta is required for submitting in Reddit, but we didn't have a chapta for submitting before either, so adding a chapta seems orthogonal to the instant username verification.

[the remark this answers is not in the document]
Comment (tsenart) [reply cmd/frontend/internal/auth/userpasswd/handlers.go]: We should not allow username or email enumeration.
